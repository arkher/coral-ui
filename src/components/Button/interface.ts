import { ReactNode } from 'react';
import { BoxProps, TextProps } from '@shopify/restyle';
import { Theme } from '../../themes/institucional';

type ButtonProps = {
  variant?: 'primary' | 'secondary' | 'tertiary';
  disabled?: boolean;
  icon?: string;
  children?: ReactNode;
  onPress: () => void;
  loading?: boolean;
  textProps?: TextProps<Theme>;
  bw?: Custom.BorderWidth;
  br?: Custom.BorderRadius;
  op?: Custom.Opacity;
  sof?: Custom.Shadow;
  sr?: Custom.Shadow;
  sop?: Custom.Shadow;
} & Partial<BoxProps<Theme>>;

type ColorsOptions = {
  [index: string]: keyof Theme['colors'];
};

type BorderWidthOptions = {
  [index: string]: Custom.BorderWidth;
};

export { ButtonProps, ColorsOptions, BorderWidthOptions };
