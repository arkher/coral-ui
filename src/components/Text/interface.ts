import { TextProps } from '@shopify/restyle';
import { ReactNode } from 'markdown-to-jsx/node_modules/@types/react';
import { Theme } from '../../themes/default';

type CustomTextProps = Partial<{
  children?: ReactNode;
  fs?: CustomFontSize;
  lh?: CustomLineHeight;
  ls?: CustomLetterSpacing;
}> &
  Partial<TextProps<Theme>>;

export { CustomTextProps };
