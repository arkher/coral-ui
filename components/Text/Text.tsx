import { createText } from "@shopify/restyle";
import { Theme } from "../../themes/default";

const Text = createText<Theme>();

export default Text;
