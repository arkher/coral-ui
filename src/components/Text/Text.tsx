import React from 'react';
import { createRestyleComponent, createText } from '@shopify/restyle';
import {
  fontSize,
  letterSpacing,
  lineHeight,
  Theme,
} from '../../themes/default';
import { CustomTextProps } from './interface';

const Text = createText<Theme>();

const CustomText: React.FC<CustomTextProps> = ({ children, ...props }) => (
  <Text variant="regular" {...props}>
    {children}
  </Text>
);

export default createRestyleComponent<CustomTextProps, Theme>(
  [fontSize, lineHeight, letterSpacing],
  CustomText,
);
