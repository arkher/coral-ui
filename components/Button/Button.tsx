import {
  BoxProps,
  createRestyleComponent,
  TextProps,
  createBox,
} from '@shopify/restyle';
import React, { ReactNode } from 'react';
import { ActivityIndicator, TouchableHighlight, View } from 'react-native';
import {
  borderWidth,
  opacity,
  shadowOffset,
  shadowOpacity,
  shadowRadius,
  Theme,
} from '../../themes/default';
import Text from '../Text/Text';

export type ButtonProps = {
  children?: ReactNode;
  onPress: () => void;
  loading?: boolean;
  textProps?: TextProps<Theme>;
  bw?: CustomBorderWidth;
  op?: CustomOpacity;
  sof?: CustomShadow;
  sr?: CustomShadow;
  sop?: CustomShadow;
} & Partial<BoxProps<Theme>>;

const SimpleBox = createBox<Theme>();

const Button: React.FC<ButtonProps> = ({
  children,
  onPress,
  loading,
  textProps,
  ...props
}) => (
  <TouchableHighlight underlayColor="transparent" onPress={onPress}>
    <SimpleBox
      py="md"
      px="md"
      backgroundColor="primaryBase"
      borderRadius="sm"
      shadowColor="black"
      {...props}
    >
      {loading ? (
        <ActivityIndicator color="white" style={{ height: 16 }} />
      ) : (
        <Text color="white" {...textProps}>
          {children}
        </Text>
      )}
    </SimpleBox>
  </TouchableHighlight>
);

export default createRestyleComponent<ButtonProps, Theme>(
  [borderWidth, opacity, shadowOffset, shadowRadius, shadowOpacity],
  Button,
);
