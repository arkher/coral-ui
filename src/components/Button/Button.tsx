import React from 'react';
import { ActivityIndicator, Pressable } from 'react-native';
import { createRestyleComponent, useTheme } from '@shopify/restyle';
import Icon from 'react-native-vector-icons/MaterialCommunityIcons';

import {
  opacity,
  shadowOffset,
  shadowOpacity,
  shadowRadius,
  Theme,
} from '../../themes';

import Text from '../Text/Text';
import Box from '../Box/Box';
import { BorderWidthOptions, ButtonProps, ColorsOptions } from './interface';

const Button: React.FC<ButtonProps> = ({
  children,
  onPress,
  loading,
  textProps,
  variant = 'primary',
  disabled,
  icon,
  ...props
}) => {
  const { colors, borderRadii } = useTheme<Theme>();

  const variantBgColor: ColorsOptions = {
    primary: !disabled ? 'primary-base' : 'neutral-lightest',
    secondary: !disabled ? 'transparent' : 'neutral-lightest',
    tertiary: 'transparent',
  };

  const variantPressedBgColor = {
    primary: colors['primary-dark'],
    secondary: colors.white,
    tertiary: colors.white,
  };

  const variantBorderColor: ColorsOptions = {
    primary: !disabled ? 'primary-base' : 'transparent',
    secondary: !disabled ? 'primary-base' : 'transparent',
    tertiary: 'transparent',
  };

  const variantBorderWidth: BorderWidthOptions = {
    primary: 'none',
    secondary: 'xs',
    tertiary: 'none',
  };

  const variantColor: ColorsOptions = {
    primary: !disabled ? 'white' : 'neutral-base',
    secondary: !disabled ? 'primary-base' : 'neutral-base',
    tertiary: !disabled ? 'primary-base' : 'neutral-base',
  };

  const variantIconColor = {
    primary: !disabled ? colors.white : colors['neutral-base'],
    secondary: !disabled ? colors['primary-base'] : colors['neutral-base'],
    tertiary: !disabled ? colors['primary-base'] : colors['neutral-base'],
  };

  const variantLoadingColor = {
    primary: colors.white,
    secondary: colors['primary-base'],
    tertiary: colors['primary-base'],
  };

  return (
    <Box
      backgroundColor={variantBgColor[variant]}
      borderColor={variantBorderColor[variant]}
      bw={variantBorderWidth[variant]}
      borderRadius="nano"
      shadowColor="black"
      height={{ phone: 48, tablet: 48 }}
      alignItems="center"
      justifyContent="center"
      flexShrink={1}
      testID="button-box"
      {...props}
    >
      <Pressable
        android_disableSound={loading}
        onPress={() => (!loading ? onPress() : () => undefined)}
        disabled={disabled}
        testID="ds-button"
        style={({ pressed }) => [
          {
            flex: 1,
            alignItems: 'center',
            justifyContent: 'center',
            width: '100%',
            height: '100%',
            opacity: disabled ? 0.5 : 1,
            borderRadius: borderRadii.nano,
          },
          {
            backgroundColor:
              pressed && !loading
                ? variantPressedBgColor[variant]
                : 'transparent',
          },
        ]}
      >
        {({ pressed }) =>
          loading ? (
            <ActivityIndicator size={16} color={variantLoadingColor[variant]} />
          ) : (
            <Box
              flexDirection="row"
              alignItems="center"
              justifyContent="center"
            >
              {icon && (
                <Box mr="quark">
                  <Icon
                    name={icon}
                    size={24}
                    color={variantIconColor[variant]}
                  />
                </Box>
              )}
              <Text
                fontWeight="600"
                fs="md"
                color={pressed ? 'primary-light' : variantColor[variant]}
                {...textProps}
              >
                {children}
              </Text>
            </Box>
          )
        }
      </Pressable>
    </Box>
  );
};

export default createRestyleComponent<ButtonProps, Theme>(
  [opacity, shadowOffset, shadowRadius, shadowOpacity],
  Button,
);
