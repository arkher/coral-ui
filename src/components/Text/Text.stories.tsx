import React from 'react';
import { storiesOf } from '@storybook/react-native';
import Text from './Text';

export default {
  title: 'Text',
  component: Text,
};

// Stories
export const BasicText = (): React.ReactNode => <Text m="sm">Ol√° mundo!</Text>;

// Add all stories to RN/Expo storybook
storiesOf('Text', module).add('Basic', BasicText);
