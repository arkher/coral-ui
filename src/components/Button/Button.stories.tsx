/* eslint-disable arrow-body-style */
import React from 'react';
import { action } from '@storybook/addon-actions';
import { withDesign } from 'storybook-addon-designs';
import { View } from 'react-native';
import Text from '../Text/Text';
import Button from './Button';
import { COLORS } from '../../themes/tokens';

type Props = {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  bg: any;
};

export default {
  title: 'Componente/Button',
  component: Button,
  argTypes: {
    bg: {
      options: COLORS,
      control: { type: 'select' },
    },
  },
  decorators: [withDesign],
};

const urlHandoff =
  'https://www.figma.com/file/3raVfIADTUZCzFOOaQ9PMQ/HANDOFF-%7C-Core-Components-Mobile-%7C-Institucional?node-id=203%3A1636';

export const Primary = ({ bg }: Props): React.ReactNode => {
  return (
    <View style={{ width: 360, marginTop: 8, marginLeft: 8 }}>
      <Button onPress={action('clicked')} bg={bg || 'primary-base'}>
        <Text>Button text</Text>
      </Button>
    </View>
  );
};

Primary.parameters = {
  design: {
    type: 'figma',
    url: urlHandoff,
  },
};

export const Secondary = (): React.ReactNode => {
  return (
    <View style={{ width: 360, marginTop: 8, marginLeft: 8 }}>
      <Button
        bg="neutral-lightest"
        borderColor="primary-base"
        bw="xs"
        onPress={action('clicked')}
      >
        <Text color="primary-base">Button text</Text>
      </Button>
    </View>
  );
};

Secondary.parameters = {
  design: {
    type: 'figma',
    url: urlHandoff,
  },
};

export const Tertiary = (): React.ReactNode => {
  return (
    <View style={{ width: 360, marginTop: 8, marginLeft: 8 }}>
      <Button bg="neutral-lightest" onPress={action('clicked')}>
        <Text color="primary-base">Button text</Text>
      </Button>
    </View>
  );
};

Tertiary.parameters = {
  design: {
    type: 'figma',
    url: urlHandoff,
  },
};

export const ButtonLoading = (): React.ReactNode => {
  return (
    <View style={{ width: 360, marginTop: 8, marginLeft: 8 }}>
      <Button
        onPress={action('loading')}
        loading
        bg="primary-base"
        alignItems="center"
      />
    </View>
  );
};
