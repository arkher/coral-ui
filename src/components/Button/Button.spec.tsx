import React from 'react';
import { cleanup, render } from '@testing-library/react-native';
import { ThemeProvider } from '@shopify/restyle';
import Button from './Button';
import Text from '../Text';
import themeMaestro from '../../themes/maestro';

afterEach(cleanup);

jest.mock('react-native-vector-icons/MaterialCommunityIcons', () => 'Icon');

test('should Button render correctly', async () => {
  const { toJSON } = render(
    <ThemeProvider theme={themeMaestro}>
      <Button variant="primary" onPress={() => undefined}>
        <Text>Label Button</Text>
      </Button>
    </ThemeProvider>,
  );

  expect(toJSON()).toMatchSnapshot();
});
