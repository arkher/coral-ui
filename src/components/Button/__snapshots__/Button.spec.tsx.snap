// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should Button render correctly 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "backgroundColor": "#0A7E19",
        "borderColor": "#0A7E19",
        "borderRadius": 4,
        "display": "flex",
        "flexShrink": 1,
        "height": 48,
        "justifyContent": "center",
        "shadowColor": "#000000",
      },
      Array [
        Object {
          "borderWidth": 0,
        },
        Array [
          Object {},
        ],
      ],
    ]
  }
  testID="button-box"
>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Array [
        Object {
          "alignItems": "center",
          "borderRadius": 4,
          "flex": 1,
          "height": "100%",
          "justifyContent": "center",
          "opacity": 1,
          "width": "100%",
        },
        Object {
          "backgroundColor": "transparent",
        },
      ]
    }
    testID="ds-button"
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "display": "flex",
            "flexDirection": "row",
            "justifyContent": "center",
          },
          Array [
            Object {},
          ],
        ]
      }
      testID="Box"
    >
      <Text
        style={
          Array [
            Object {
              "color": "#FFFFFF",
              "display": "flex",
              "fontFamily": "\\"Inter\\", sans-serif",
              "fontWeight": "600",
            },
            Array [
              Object {
                "fontSize": 14,
              },
            ],
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "display": "flex",
                "fontFamily": "\\"Inter\\", sans-serif",
                "fontWeight": "400",
              },
              Array [
                Object {},
              ],
            ]
          }
        >
          Label Button
        </Text>
      </Text>
    </View>
  </View>
</View>
`;
